Added Longjump interop limitations documentation. (#16703)
dotnet/docs: docs/standard/native-interop/exceptions-interoperability.md

---
title: "Exceptions Interoperability"
ms.date: "01/16/2020"
ms.technology: dotnet-standard
helpviewer_keywords:
  - "unmanaged code, exceptions"
  - "exceptions, unmanaged code"
  - "interop, exceptions"
  - "exceptions, interop"
---
# Working with Interop Exceptions in Unmanaged Code

> Unmanaged code exception interop is not supported in non-Windows platforms, due to the portability issues that arise, since managed code can't know how these exception mechanisms work under the covers. The Unix ABI has no definition for this and therefore, exceptions end up resulting in unpredictable behaviors and potential crashes.

Reviewer: I would also explicitly mention: `Unmanaged code exception interop is supported on Windows platforms only.`

Author: Ok! Would this be added, or replacing `not supported in non-Windows platforms`? I ask because of the potential redundancy.

Reviewer: It should be added.

Reviewer: The doc should be explicit about both the behavior on Windows and the behavior on non-Windows.

## Setjmp/Longjmp Behaviors

> `Setjmp`/`Longjmp` interop is held to the same constraints as exception handling as it is not supported to skip over managed frames this way.

Reviewer: ```suggestion
The interop with `setjmp`/`longjmp` C functions is not supported. It is not supported to use `longjmp` to skip over managed frames.
```

Reviewer: Could you please apply this suggestion as well?

Author: Sorry Jan, for some reason, it didn't show up in the email and I overlooked it when reviewing here. Let me do it.

---
title: "Exceptions Interoperability"
ms.date: "01/16/2020"
ms.technology: dotnet-standard
helpviewer_keywords:
  - "unmanaged code, exceptions"
  - "exceptions, unmanaged code"
  - "interop, exceptions"
  - "exceptions, interop"
---

> # Working with Interop Exceptions in Unmanaged Code

Reviewer: We do support interop for exceptions on Windows. It is well-tested area and it is fully supported, e.g. as part of managed C++.

We do not support interop for exceptions on non-Windows.

We should declare setjmp/longjmp as unsupported everywhere. It is not tested well, and it is unlikely that it works 100% even though it happens to work for simple cases.

I am sorry about the confusion ... hope that this makes it clear.

Author: Let me see if I got this correctly. In the first section, `Working with Interop Exceptions in Unmanaged Code`, I can add the Windows support at the beginning, making the line look something like _"Unmanaged code exception interop is not supported in non-Windows platforms, due to..."_.

Then, rephrase the second section to remove the section that states that on Windows there is support to unwind the stack and deal with external exceptions. Leave the Unix ABI not supporting section as is, as well as the unpredictable behaviors and potential crashes section.

Something like this?

Reviewer: > Leave the Unix ABI not supporting section as is

Yes, but it should be moved into the "Exceptions chapter".

The section about setjmp/longjump should just say that it is not supported. It does need to explain in detail why it is not supported.
